src/prop/theory_proxy.h

```diff
@@ -87,7 +87,12 @@ class TheoryProxy : protected EnvObj, public Registrar
 
   void theoryCheck(theory::Theory::Effort effort);
 
+  /** Get an explanation for literal `l` and save it on clause `explanation`. */
   void explainPropagation(SatLiteral l, SatClause& explanation);
+  /** Notify that current propagation inserted at lower level than current. */
```

Comment:
These will be called by the SAT solver in three scenarios: explanation of propagation is added at lower level; asserted clause is added at lower level (either as soon as the assertion is requested as when the clause, which is a lemma, is added later on after the SAT solver is no longer busy).

What these calls do is invoke similar methods in proof cnf stream which will track the proofs of these optimized clauses so that we don't lose them in an eventual pop.

We can only know that a clause is optimized in the SAT solver *after* it is asserted, so since the information has to be tracked outside (in the case of the proofs that are in the CNF stream) I think this setup is the more natural one.

Reply:
Ok, can you add this to the documentation here?

Also, I'd prefer these names were a bit more specific, e.g. `notifyCurrentPropagationWasOpt` or something like this.  Ditto for the other method.

Reply:
Added, PTAL